### ISSUE TITLE
Slower on GPU

### ISSUE THREAD

**Issue by davidsohutskay** [sciann/sciann#8]
Is this normal? I tried comparing CPU with GPU performance, I get about 1 epoch per 4s on CPU and 1 epoch per 16s on GPU.

Everything looks like it is running on GPU but it is quite a bit slower:

![image](https://user-images.githubusercontent.com/59542631/96388620-17531280-1178-11eb-9398-61404f708072.png)

Versus:

![image](https://user-images.githubusercontent.com/59542631/96388734-cbed3400-1178-11eb-9f42-752361670b03.png)

I am running on:

![image](https://user-images.githubusercontent.com/59542631/96388719-aceea200-1178-11eb-9548-f20951d9ae1c.png)

The network is a 10*[24] FCNN with 3 inputs and 3 outputs.

Any ideas? Is this normal behavior? In addition, are there any benchmark programs that I could run and compare timing?

Thanks for the help! My results look really nice so far but it would always be nice to be faster.

**Comment by ehsanhaghighat** [comment]
you should get the same performance as you were to use pure Keras code. If the network is small and/or if the batch size is too small, the overhead from graph-based computations dominates the performance. In that case, I usually get better performance with CPU. Do you have some adjustments to the code to support CNNs? If you can, your contribution to the repository is welcomed ...

**Comment by davidsohutskay** [comment]
Ah sorry about that, just an atypical abbreviation - I meant a "fully-connected neural network". A convolutional network would be interesting though.

Interestingly, I do see a pretty big speedup with DeepXDE (about 10x faster). I think the settings are about the same, a 12*[32] network, 32 or 64 batch size, roughly 15000 sample points. I will do a bit more investigation. I do seem to get better convergence with SciANN for some reason. And I like the reduce LR on plateau.

**Comment by ehsanhaghighat** [comment]
Interesting... that's a deep network and I would expect that gpu should provide speed up.

**Comment by mariusmerkle** [comment]
Hi @davidsohutskay and @ehsanhaghighat !

I use both SciANN and DeepXDE libraries as well and, for now, experience the same differences:

- SciANN yields better accuracy. I simulate the time-dependent 2D heat equation and achieve low relative L2 errors that I couldn't achieve with any hyper parameter configuration in DeepXDE.
- DeepXDE is significantly faster. I didn't measure the difference yet, but when running the same problem setup (same PDE, network architecture, number of domain points etc.), each epoch takes much longer in SciANN.

Did you find out anything else? @davidsohutskay

**Comment by davidsohutskay** [comment]
Hi @mariusmerkle, I am not sure. I never really figured out why the speed difference. I am still using SciANN and still get better convergence than deepXDE like you said, but I have just been running things on multithreaded CPU. Despite the slower epochs the better convergence is worth it and the program still finishes in a reasonable amount of time for me.

**Comment by ZPLai** [comment]
> Hi @davidsohutskay and @ehsanhaghighat !
> 
> I use both SciANN and DeepXDE libraries as well and, for now, experience the same differences:
> 
> * SciANN yields better accuracy. I simulate the time-dependent 2D heat equation and achieve low relative L2 errors that I couldn't achieve with any hyper parameter configuration in DeepXDE.
> * DeepXDE is significantly faster. I didn't measure the difference yet, but when running the same problem setup (same PDE, network architecture, number of domain points etc.), each epoch takes much longer in SciANN.
> 
> Did you find out anything else? @davidsohutskay

Hi @mariusmerkle I used deepXDE and sciANN both as well, and when comparing their performances on CPU and GPU, I have similar experience with you.
